src/charm.py

```diff
@@ -1,117 +1,278 @@
+            event.fail("Certificates relation not created.")
+            return
 
-    def _on_httpbin_pebble_ready(self, event: ops.PebbleReadyEvent) -> None:
-        """Define and start a workload using the Pebble API.
+        tls_rel_data = tls_certificates_relation.data[self.app]
+        if not tls_rel_data.get(f"certificate-{hostname}"):
+            event.fail("Certificate not available")
+            return
 
-        Change this example to suit your needs. You'll need to specify the right entrypoint and
-        environment configuration for your specific workload.
+        event.set_results(
+            {
+                f"certificate-{hostname}": tls_rel_data[f"certificate-{hostname}"],
+                f"ca-{hostname}": tls_rel_data[f"ca-{hostname}"],
+                f"chain-{hostname}": tls_rel_data[f"chain-{hostname}"],
+            }
+        )
 
-        Learn more about interacting with Pebble at at https://juju.is/docs/sdk/pebble.
+    def _on_certificates_relation_created(self, event: RelationCreatedEvent) -> None:
+        """Handle the TLS Certificate relation created event.
 
         Args:
-            event: event triggering the handler.
+            event: The event that fires this method.
         """
-        # Get a reference the container attribute on the PebbleReadyEvent
-        container = event.workload
-        # Add initial Pebble config layer using the Pebble API
-        container.add_layer("httpbin", self._pebble_layer, combine=True)
-        # Make Pebble reevaluate its plan, ensuring any services are started if enabled.
-        container.replan()
-        # Learn more about statuses in the SDK docs:
-        # https://juju.is/docs/sdk/constructs#heading--statuses
-        self.unit.status = ops.ActiveStatus()
+        if not self._are_relations_ready():
+            self.unit.status = WaitingStatus("Waiting for certificates relation to be created")
+            event.defer()
+            return
+        hostname = self.get_hostname()
+        if not hostname:
+            self.unit.status = BlockedStatus("Waiting for hostname to be configured")
+            event.defer()
+            return
+        self._tls.certificate_relation_created(hostname)
 
-    def _on_config_changed(self, event: ops.ConfigChangedEvent) -> None:
-        """Handle changed configuration.
+    def _on_certificates_relation_joined(self, event: RelationJoinedEvent) -> None:
+        """Handle the TLS Certificate relation joined event.
 
-        Change this example to suit your needs. If you don't need to handle config, you can remove
-        this method.
+        Args:
+            event: The event that fires this method.
+        """
+        if not self._are_relations_ready():
+            self.unit.status = WaitingStatus("Waiting for certificates relation to be created")
+            event.defer()
+            return
+        hostname = self.get_hostname()
+        if not hostname:
```

Comment:
I believe this could be for a future PR but using charm_state + validator could make it easier to guarantee that hostname is a required configuration. Synapse has the same case for server_name.

Reply:
good point, I'll add it to a followup PR.